pkg/scheduler/scheduler.go

```diff
@@ -626,7 +626,11 @@ func (s *Scheduler) requeueAndUpdate(ctx context.Context, e entry) {
 	log.V(2).Info("Workload re-queued", "workload", klog.KObj(e.Obj), "clusterQueue", klog.KRef("", e.ClusterQueue), "queue", klog.KRef(e.Obj.Namespace, e.Obj.Spec.QueueName), "requeueReason", e.requeueReason, "added", added, "status", e.status)
 
 	if e.status == notNominated || e.status == skipped {
-		if workload.UnsetQuotaReservationWithCondition(e.Obj, "Pending", e.inadmissibleMsg) {
+		s.queues.Lock()
+		updated := workload.UnsetQuotaReservationWithCondition(e.Obj, "Pending", e.inadmissibleMsg)
+		s.queues.Unlock()
```

Comment:
Oh this was pretty bad.
But we shouldn't be locking the queue.
Instead, we should be copying the Workload and modifying a copy.
Or expose the `workload.admissionPatch` function (perhaps with a better name) to copy the minimal information to be send via SSA.

Reply:
We need the side effects of `workload.UnsetQuotaReservationWithCondition(e.Obj,...`  so for additional calls for the same workload `UnsetQuotaReservationWithCondition` returns false and  avoid unnecessary API updates.

Reply:
That will come through the event handlers anyways.

But another alternative is to update the Workload object before calling `RequeueWorkload`. This prevents locking twice.
`ApplyAdmissionStatus` can still be done after.

Reply:
> But another alternative is to update the Workload object before calling `RequeueWorkload`. This prevents locking twice.
> `ApplyAdmissionStatus` can still be done after.

this might be the best option

Reply:
> That will come through the event handlers anyways.

Should we send event only on update too?

Reply:
No, the event should be reported every time there is an attempt.

The status should only be updated once.

Reply:
The "event handlers" process object updates, no new events. I know, it's confusing :)

Reply:
> But another alternative is to update the Workload object before calling `RequeueWorkload`. This prevents locking twice.

`RequeueWorkload` replacing `e.Obj` on line 354

https://github.com/kubernetes-sigs/kueue/blob/4a151eec00a9b374a82dd8e9193269ea117c125c/pkg/queue/manager.go#L338-L355

and also not fixing race condition problem.

Reply:
If `RequeueWorkload` getting last version of `e.Obj` and setting it to `Info` object we can use `DeepCopy()` after `RequeueWorkload `.

Reply:
Right, so if you update the object before calling RequeueWorkload, then it would have the up-to-date information when you enter the function.

Reply:
In this case we don't need to Lock queue twice and it's fixing race condition problem.

Reply:
> Yes, but cloning a big object such as a workload isn't exactly cheap. It
looks like we can avoid it, so I prefer that.

Agree, but on this case we don't have actual information about Workload because  `e entry` coming as an object (not a reference) and all changes will be lost after end of this function.

Reply:
Ok, I went through all the logic to understand the origin of the pointer that gets the race access.

Certainly cloning the object is the safest solution and it preserves the behavior. However, I think we can do better.
ApplyAdmissionStatus already produces a minimal version of the status to send via SSA. We can add a function `AdmissionStatusPatch`, and just add the condition there, and then call another function `ApplyAdmissionStatusPatch`.

Reply:
Done